arch/posix/CMakeLists.txt

```diff
@@ -1,12 +1,53 @@
 # SPDX-License-Identifier: Apache-2.0
 
-if (CONFIG_64BIT)
-	# some gcc versions fail to build without -fPIC
-	zephyr_compile_options(-m64 -fPIC)
-	zephyr_link_libraries(-m64)
-else ()
-	zephyr_compile_options(-m32)
-	zephyr_link_libraries(-m32)
+# @Intent: Detect host architecture (see 'man 1 arch')
+execute_process(
+  COMMAND
+  arch
+  OUTPUT_VARIABLE
+  BUILD_ARCH
+  OUTPUT_STRIP_TRAILING_WHITESPACE
+)
```

Comment:
This will not work on windows.
Is that acceptable ?
After all we are using a cross compiler, so I assume it should be possible to build on windows.

Reply:
Is `native_posix` supported under windows to begin with?

Reply:
Also, on the machine I was testing on, I originally was cross-compiling from aarch64-linux-gnu to arm-non-eabi :-) I know it sounds masochistic but really it's not that bad.

Reply:
@tejlmand native_posix is meant to be used to compile targeting your host, so you can test and debug in your own workstation. It uses the host libC, and the host kernel APIs (it requires the kernel offering POSIX threads and other POSIX goodies).
Today I do not think it is really possible to crosscompile from Windows, and would not be too useful to do so (It is a quite bizarre use case, like people developing Linux apps in Windows, in which case they probably end up using WSL or so, which should then work).
Crosscompiling from one Linux platform to another would still be possible, but I would see it as a niche (e.g. using cheaper HW for multidevice tests yet wanting a Linux host below Zephyr).
In principle the POSIX arch can target any POSIX compatible OS, but in reality it is only working on Linux so far (and so far only x86).
Running natively on the Windows kernel is not possible at all or intended by design.

Reply:
@aescolar I know the target `native_posix` is for local host compilation (and I'm not sure that will work on windows in first place, unless you use WSL, cygwin, or similar), I just got tricked by @cfriedt setting cross compiler at first (which is why I edited the comment).

But the question still remains valid, do we want to call out to host commands that may not be available on a system that we are supporting without some safeguard ?

People tend to copy-paste, and suddenly such code could show up in other parts of the build system, breaking it for windows users.
Therefore I would prefer if there were a generic way for doing this test, or ensure a `find_program()` is used, so that proper error message can be printed to user if running on a system where `arch` for some reason is not available.

Reply:
Fair enough. I guess it could try call arch and if it cant throw a warning but assume x86_64 which is the most typical case

Reply:
as minimum:
```
find_program(ARCH_EXECUTABLE arch)

if(${ARCH_EXECUTABLE} STREQUAL ARCH_EXECUTABLE-NOTFOUND)
  # Error out or warning ?
else()
  execute_process(
    COMMAND
    ${ARCH_EXECUTABLE}
    OUTPUT_VARIABLE
    BUILD_ARCH
    OUTPUT_STRIP_TRAILING_WHITESPACE
  )
endif()
```

Reply:
If it's easier to digest, we can use the POSIX-required `uname -m`.

Reply:
> Executable should exist on each platform that is POSIX-compiliant.

which windows is not.

As the Zephyr build system is also supporting windows, then there are a couple of reason why calls to host tools should be done with care.
- The tool may not be available, which is why `find_program` is used, and that goes in general.
  `west`, `dtc`, `c compiler`, etc. are all looked up using `find_program`
- Users may see this code, and consider this a nice way to do such testing, and copy it into other parts of the build system, and then refer: `But it's done like that in xyz.cmake`.

The support for multiple OS'es is also why `python` is used, and not `sed`, `awk`, and similar tools.

And then finally for my own sake.
Whenever I review CMake code, I prefer not to consider whether this part of CMake code might be executed on windows or not.

Reply:
A possibility would also be to extend your python script:
`python -c 'import platform; print(platform.machine())'
returns `x86_64` on Linux, AMD64 on windows 64 bit.

Unsure what 32 bit systems return as the Python docs are a bit brief on this: https://docs.python.org/3.8/library/platform.html#platform.machine

but for this small task, using `find_program(ARCH ...)` and then `if(${ARCH_EXECUTABLE} STREQUAL ARCH_EXECUTABLE-NOTFOUND)` the simplest approach.

Reply:
Nice - that works really well. Also, found `CMAKE_HOST_SYSTEM_NAME`. 

I wonder if it would make sense to split things out into `CMAKE_HOST_SYSTEM_NAME`.`CMAKE_HOST_SYSTEM_PROCESSOR`.cmake and then just include the one file (if it exists? is that possible?).

I'm not a CMake expert by any means, so thanks for the tips :-)